tools/ChangelogGenerator/changelog_generator.py

```diff
@@ -17,6 +17,89 @@
 contributors = {}
 raw_commits_info = {}
 
+def print_commits(commits: dict, indent: str = "") -> (str, list):
+    if not commits: return ("", [])
+    ret_message = ""
+    ret_contributor = []
+
+    for commit_hash, commit_info in commits.items():
+        commit_message = commit_info["message"]
+
+        if not with_commitizen:
+            commitizens = r"(?:build|chore|ci|docs?|feat|fix|perf|refactor|rft|style|test)"
+            commit_message = re.sub(rf"^(?:{commitizens}, *)*{commitizens} *(?:\([^\)]*\))*: *", "", commit_message)
+    
+        ret_message += indent + "- " + commit_message
+    
+        mes, ctrs = print_commits(commit_info["branch"], indent + "   ")
+    
+        if not ignore_merge_author or not commit_info["branch"]:
+            author = commit_info["author"]
+            if author not in ctrs: ctrs.append(author)
+            if committer_is_author:
+                committer = commit_info["committer"]
+                if committer not in ctrs: ctrs.append(committer)
+    
+        for ctr in ctrs:
+            if ctr == "web-flow": continue # 这个账号是 GitHub 在 Merge PR 时的 committer
+            if ret_contributor.count(ctr) == 0:
+                ret_contributor.append(ctr)
+            ret_message += " @" + ctr
+    
+        if with_hash:
+            ret_message += f" ({commit_hash})"
+    
+        ret_message += "\n" + mes
+        
+    return ret_message, ret_contributor
+
+def updateCommits(commit_message, sorted_commits, update_dict):
+    translations = {
+                '修复': 'fix', 
+                '新增': 'feat',
+                '改进': 'perf', 
+                '更新': 'perf', 
+                '优化': 'perf', 
+                '重构': 'perf',
+                '其他': 'other'
+            }
+    oper = 'other'
+    for key, trans in translations.items():
+        if key in commit_message or commit_message.startswith(trans):
+            oper = trans
+            break
+    sorted_commits[oper].update(update_dict)
```

Comment:
这里没啥大问题，但是和原来的不完全等价，因为原来是先匹配一遍关键词，再匹配一遍英文的开头。对于`feat: 更新`这样的 commit，原来会分类成`pref`而这里会分类成`feat`，我不知道那个比较好。但是我觉得两个不要混在一起匹配比较好，这样改优先级也比较方便。

```suggestion
KEYWORD_TO_COMMIT_TYPE = {
    '修复': 'fix',
    '新增': 'feat',
    '改进': 'perf',
    '更新': 'perf',
    '优化': 'perf',
    '重构': 'perf',
    '其他': 'other'
}

COMMIT_TYPES = ['fix', 'feat', 'perf']

def updateCommits(commit_message, sorted_commits, update_dict):
    commit_type = 'other'
    for keyword, type_ in KEYWORD_TO_COMMIT_TYPE.items():
        if keyword in commit_message:
            commit_type = type_
            return
    for type_ in COMMIT_TYPES:
        if commit_message.startswith(type_):
            commit_type = type_
            return
    sorted_commits[commit_type].update(update_dict)
```

Reply:
已修正